### ISSUE TITLE
Streamline transifex and translations

### ISSUE THREAD

**Issue by Cqoicebordel** [searx/searx#238]
I'm playing with Transifex and the whole translation process, so don't be alarmed by anything you can see out there.

There are a few issue I want to iron out before submitting anything. But I think I will have something tonight.

I'll post it here.

**Comment by Cqoicebordel** [comment]
Ok. It's a mess : 
There are differents translations on transifex, for the same language. Look at Russian and Russian (Russia). It created two folders too in the searx tree, ru and ru_RU.

So @asciimoo can you delete the languages on transifiex ru_RU, el_GR, and pt_BR ? It will simplify a bit the scripting part.
And when adding a language, we should **_only**_ use the languages with a two or three chars code, not the full unicode ones.

For now, I can fully automatically update transifex daily from the code. I'm left with the other way around (transifex -> code).

**Comment by Cqoicebordel** [comment]
There is the script. Be careful, it doesn't work propelly yet, because of the languages to be deleted in Transifex, and because it doesn't create a git branch before commiting (should it ?)

Anyway, have a look :

``` bash
# Get latest updates from upstream
git fetch --progress --prune upstream
# And push it to origin
git push --porcelain --progress --recurse-submodules=check origin refs/remotes/upstream/*:refs/heads/*
# Fast forward the local git repository
git merge --no-commit --ff origin/master

# Find every translatable string in the updated code
./utils/update-translations.sh
# And compile the translations (Transifex uses the compiled files)
pybabel compile -d searx/translations/
# Push to Transifex the new strings to be translated, and the new translations from the code
tx push -s -t
# Get from Transifex the latest updated translations
tx pull -a

# Uncomment this line to debug the script. This one is a life savior.
#set -xv

# This whole block serve only one purpose : dismiss translations files that has only one change :
# the date at the beginning of the file (because of pybabel)
find -path './searx/translations/*/LC_MESSAGES/messages.po' | while read d; do 
    git diff --stat "$d" | grep changed | while read -a arr
    do
        insertion="${arr[3]}"
        deletion="${arr[5]}"
        if [ $insertion == "1" ]
        then
            if [ $deletion == "1" ]
            then
                files=`echo $d | head -c -3`
                git checkout -- $files*
            fi
        fi
    done
done

# Compile the new translations including those from Transifex
pybabel compile -d searx/translations/

# Add the folder of translations to the commit
git add './searx/translations/'
# And commit it
git commit -m "Translations update\nFrom Transifex (with script)"
# And finally, push it to origin
git push
```

It works globally. I couldn't test the case there isn't any change in a tranlation file, because right now, all of them have some...  
You have to have configured Tx first, and I use a different folder for the git repository, so there are no other changes in it. Also, I used SSH to connect to github with git, and using a private key, so Github doesn't ask for my credentials every time.
When the supperfluous languages and the branch issue will be dealt with, I will commit it in `utils`.  
And when that is done, I will be able to have a cronjob running it daily (or whenever).
I'm also fiddling with something to do the pull request automatically.

A web server doing this would be better, and web hooks could be used, but I don't have a web facing server, and it's a start.

Any comments are welcome, of course :)

**Comment by pointhi** [comment]
I'm not sure, but I think `tx push -s -t` overwrite the translations on transifex, so `tx -a` couldn't fetch new strings (and probably overwrite new translations).

If we change it to `tx push -s` we would not overwrite translations, and updating only the source-file.

**Comment by Cqoicebordel** [comment]
Good point. But I thought that pushing to Transifex would allow to keep the old translations. But upon rereading the doc http://docs.transifex.com/developer/client/push I can say for sure that Transifex docs are not very clear ;)

My point was to be able to handle a translation coming from Github and from Transifex simultaneously.

But streamlining meaning streamlining, I would propose to delete every .po files in the repository, so the translators would/should/could only use Tx, resolving this issue. But it would also reduce the choices for the contributors. So I let @asciimoo decide.

@dimqua is there a reason you proposed your translations fixes using code (pull request) and not Transifex ?

**Comment by asciimoo** [comment]
@Cqoicebordel i don't want to depend (more) on third party services, so i'd leave the .po files in the repo. Alternatively, a contribution guide could clarify these kind of questions.

**Comment by Cqoicebordel** [comment]
In that case, we should close searx on Transifex completly.
Keeping both means two things : 
- It's a mess to do things properly
- There could be conflict in translations.

For streamlining sake, we _must_ only have one access point, at least for "tourists" translators.

**Comment by asciimoo** [comment]
Keeping the po files in the repo doesn't mean that we can't define a policy to edit translations only on transifex. It just makes easier to create local modifications.

**Comment by Cqoicebordel** [comment]
For local modifications, the po can be generated using update-translations.sh as described in the wiki. Searx only uses the mo.

**Comment by asciimoo** [comment]
update-translations.sh won't create missing po files (nor fill it with the translations stored in the corresponding .mo) afaik, only updates existing ones.

**Comment by Cqoicebordel** [comment]
Oh. True.

In that case, would you agree to remove the po files if update-translations.sh would create the po files ?
(I'll take care of doing it of course)

**Comment by asciimoo** [comment]
I just can't see the benefit of it. Let's say that we wont pull translation requests from now, isn't it enough?

**Comment by Cqoicebordel** [comment]
There is no benefit in the code front. The benefit is elsewhere :  
The core of it is in the streamlining : it's easier for everyone to contribute, maintain, deploy. You don't have to have a bunch of wiki pages to explain how to do this kind of stuff, there is only one option.

It's like the installation : For now, there are a bunch of options on how to deploy searx. That's great. But, there aren't enough doc for it, scripts for it, and everyone is a bit confused when it comes to maintaining it, or solving others issues.

So in the case of the translations, it's only to avoid confusion, difficulty of maintainability, and overall, allowing us to streamline the experience by providing tools (scripts) that do everything "unimportant" around searx.

**Comment by asciimoo** [comment]
In this case you also have to tell to the contributors that transifex is the place where the translation happens. Anyway, what about removing .po files if there will be another incident around the topic?

**Comment by Cqoicebordel** [comment]
Hum, I don't understand your question. Can you reformulate it ?

**Comment by hovancik** [comment]
Hi, so it this actively used?  https://www.transifex.com/asciimoo/searx/
I have made Slovak translation but don't know how to proceed... 

a bit of out topic: 
how can I try it locally before contributing Slovak translation? I am on mac.

**Comment by asciimoo** [comment]
@hovancik first of all, thank you the translations.
I'll update the repo with the new transifex translations.
If you want to experiment with the translations in your local instance, see http://asciimoo.github.io/searx/dev/translation.html for more information.

**Comment by hovancik** [comment]
hi, I am trying to upload to tx and I am getting: 

`tx set --auto-local -r searx.messagespo 'searx/translations/sk/LC_MESSAGES/messages.po' --source-lang en --type PO --source-file messages.pot --execute`
```
Usage: tx [tx_options] set [options] [args]

tx: error: The expression you have provided is not valid.
```
